### PR TITLE
DAOS-8040 control: enable vmd support by caching state in bdev forwarder

### PR DESCRIPTION
**Pull request by tanabarr**
VMD state is detected during NVMe storage prepare phase, cache this
state in the BdevAdminForwarder which handles the forwarding of requests
from daos_server to daos_admin process instances. The state can then be
used to determine whether or not to enable VMD processing in bdev backend.

Signed-off-by: Tom Nabarro <tom.nabarro@intel.com>

### CODE REVIEW

File: src/control/server/storage/bdev/backend_vmd.go
Comment (kjacque) [inline src/control/server/storage/bdev/backend_vmd.go:232]: I found these conditionals a little hard to understand. Here's a suggestion that could help make it clearer:

```suggestion
	if vmdReq.PCIAllowList == "" {
	        if req.PCIAllowList != "" {
	        	// all VMD devs filtered out by allow list
			log.Debugf("vmd prep: %v devices not allowed", vmdPCIAddrs)
		        return nil, nil
		}
		if req.PCIBlockList != "" {
			// all VMD devs filtered by block list
			log.Debugf("vmd prep: %v devices blocked", vmdPCIAddrs)
			return nil, nil
		}
	}
```

**Review by kjacque — COMMENTED** [review]
Good reorganization -- it's a lot cleaner to put most of the VMD logic in the backend.

File: src/control/server/storage/bdev/backend.go
Comment (sydneyvanda) [inline src/control/server/storage/bdev/backend.go:285]: this output is a little confusing:
DEBUG 00:44:16.196015 backend.go:285: check bdev cfg for engine-0: vmd detected, checking addresses (input [0000:5d:05.5], existing 5d0505:01:00.0, 5d0505:03:00.0): new [5d0505:01:00.0 5d0505:03:00.0]
maybe "new" should be "added bdevs" or something like that?

File: src/control/server/storage/bdev/backend_vmd.go
Comment (sydneyvanda) [inline src/control/server/storage/bdev/backend_vmd.go:100]: We should only grep for "Volume Management Device" now, since the 201d pci ID is only true for Skylake/Cascade Lake and has changed in ICX systems. Should simplify the logic a bit.

**Review by sydneyvanda — APPROVED** [review]
LGTM with just minor comments. Tested locally and works great. Also fixes go unit test failures I was seeing with VMD devices being present on the system.

**Comment by sydneyvanda** [comment]
@tanabarr @mjmac We should state in the commit message before PR is merged that VMD will still fail for multiple engines on the same server due to the VFIO permission issues.

**Comment by sydneyvanda** [comment]
> > @tanabarr @mjmac We should state in the commit message before PR is merged that VMD will still fail for multiple engines on the same server due to the VFIO permission issues.
> 
> @sydneyvanda: Hmm. This seems like a pretty critical flaw... Can you elaborate on this?

Part of the hotplug work that is yet to be done is to configure a hotplug filter (SPDK patch) which is a PCIe bus range for each engine on the same node. @tanabarr please correct me if I am wrong, but I believe as apart of that work we were going to handle the VFIO device permission issues that are caused with SPDK not honoring the device addresses listed in the daos_nvme.conf but instead trying to scan and probe all of the devices. I am not 100% clear on how this permission issue is currently fixed in the control plane for regular NVMe devices with the VFIO driver on multiple engines, but it definitely does not work with VMD.

Tom was planning on looking into this issue more when he got back from vacation, and I can do some more testing myself this week. However, enabling VMD is targeted for TB5 so we need to figure out if single engine use will be ok for now, or if we should hold until this gets sorted out.

**Comment by mjmac** [comment]
> Tom was planning on looking into this issue more when he got back from vacation, and I can do some more testing myself this week. However, enabling VMD is targeted for TB5 so we need to figure out if single engine use will be ok for now, or if we should hold until this gets sorted out.

From my perspective, I am OK with landing this in order to maintain progress on VMD development as long as we are not regressing non-VMD functionality. Just to be clear: The single engine restriction applies only when VMD is enabled, right? In other words, we're not going to break multi-engine support for non-VMD configurations with this patch?

As for whether or not single-engine VMD configs are OK for now, I suppose ultimately that's up to @johannlombardi.

**Comment by ravalsam** [comment]
> From my perspective, I am OK with landing this in order to maintain progress on VMD development as long as we are not regressing non-VMD functionality. Just to be clear: The single engine restriction applies only when VMD is enabled, right? In other words, we're not going to break multi-engine support for non-VMD configurations with this patch?
> 
This PR ran some tests with multi-engine with NVMe and that seems to work fine. 
- daos_core (https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-6362/20/artifact/Functional%20Hardware%20Medium/daos_test/daos_core_test.py/daos_configs/daos_server.wolf-138.yml/*view*/)
- EC smoke (https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-6362/20/artifact/Functional%20Hardware%20Large/erasurecode/ec_ior_smoke.py/daos_configs/daos_server.wolf-146.yml/*view*/)

**Comment by johannlombardi** [comment]
Yeah, that's a pretty big gap since all our configs are dual engines now. Unless this can be handled as a bug fix, we are going to have to de-scope VMD from 2.0.

**Review by johannlombardi — APPROVED** [review]

**Pull request merged**
